etcdctl: refactor the way to check cluster health

yichengq opened the pull request:
This method uses raft status exposed at /debug/varz to determine the
health of the cluster. It uses whether commit index increases to
determine the cluster health, and uses whether match index increases to
determine the member health.

This could fix the bug that fails to detect follower is unhealthy
because it doesn't rely on whether message in long-polling connection is sent.

This health check is stricter than the old one, and reflects the
situation that whether followers are healthy in the view of the leader. One
example is that if the follower is receiving the snapshot, it will turns
out to be unhealthy because it doesn't move forward.

This will reflect the healthy view in the raft level, while connectivity checks reflects the healthy view in transport level.

This is a prototype, and I am wondering whether this fits our goal.

fixes #2711

xiang90 commented:
This is what I would expect to see as the first step. I am still working on the connectivity stuff.

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,30 @@ func handleClusterHealth(c *cli.Context) {
-	fmt.Println("cluster is healthy")
-	// self is healthy
+	if rs1.Commit > rs0.Commit {
+		fmt.Println("cluster is healthy")
```
cluster is healthy: raft is making progress [%v -> %v]

xiang90 replied:
we can also print out who is the leader

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,30 @@ func handleClusterHealth(c *cli.Context) {
+	if rs1.Commit > rs0.Commit {
+		fmt.Println("cluster is healthy")
+	} else {
+		fmt.Println("cluster is unhealthy")
```
cluster is unhealthy: raft is not making progress [%v]

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,30 @@ func handleClusterHealth(c *cli.Context) {
-		if fs1.Counts.Success <= fs0.Counts.Success {
-			prints = append(prints, fmt.Sprintf("member %s is unhealthy\n", name))
+		if pr1.Match <= pr0.Match {
+			prints = append(prints, fmt.Sprintf("member %s is unhealthy\n", id))
```
can we also print out its progress?

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,30 @@ func handleClusterHealth(c *cli.Context) {
+			prints = append(prints, fmt.Sprintf("member %s is unhealthy\n", id))
 		} else {
-			prints = append(prints, fmt.Sprintf("member %s is healthy\n", name))
+			prints = append(prints, fmt.Sprintf("member %s is healthy\n", id))
```
same here

yichengq commented:
I updated the output:
```
all health:
$ ./bin/etcdctl cluster-health
cluster is healthy: raft is making progress [commit index: 12332->12334]
leader of the cluster is member 6e3bd23ae5f1eae0
member 6e3bd23ae5f1eae0 is healthy [match: 12334, next: 12335, state: ProgressStateProbe]
member 924e2e83e93f2560 is healthy [match: 12334, next: 12335, state: ProgressStateReplicate]
member a8266ecf031671f3 is healthy [match: 12334, next: 12335, state: ProgressStateReplicate]

2 out of 3 healthy:
$ ./bin/etcdctl cluster-health
cluster is healthy: raft is making progress [commit index: 12381->12383]
leader of the cluster is member 6e3bd23ae5f1eae0
member 6e3bd23ae5f1eae0 is healthy [match: 12383, next: 12384, state: ProgressStateProbe]
member 924e2e83e93f2560 is unhealthy [match: 12373, next: 12374, state: ProgressStateProbe]
member a8266ecf031671f3 is healthy [match: 12383, next: 12384, state: ProgressStateReplicate]
```

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,31 @@ func handleClusterHealth(c *cli.Context) {
+	} else {
+		fmt.Printf("cluster is unhealthy: raft is not making progress [commit index: %v]\n", rs0.Commit)
+	}
+	fmt.Printf("leader of the cluster is member %v\n", rs0.Lead)
```
sounds too verbose. how about leader is ?

yichengq replied:
ok

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,31 @@ func handleClusterHealth(c *cli.Context) {
-		if fs1.Counts.Success <= fs0.Counts.Success {
-			prints = append(prints, fmt.Sprintf("member %s is unhealthy\n", name))
+		if pr1.Match <= pr0.Match {
+			prints = append(prints, fmt.Sprintf("member %s is unhealthy [match: %v, next: %v, state: %v]\n", id, pr1.Match, pr1.Next, pr1.State))
```
i think we should tell user why do we think this member is unhealthy.

yichengq replied:
ok

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,31 @@ func handleClusterHealth(c *cli.Context) {
+			prints = append(prints, fmt.Sprintf("member %s is unhealthy [match: %v, next: %v, state: %v]\n", id, pr1.Match, pr1.Next, pr1.State))
 		} else {
-			prints = append(prints, fmt.Sprintf("member %s is healthy\n", name))
+			prints = append(prints, fmt.Sprintf("member %s is healthy [match: %v, next: %v, state: %v]\n", id, pr1.Match, pr1.Next, pr1.State))
```
well... can we should the progress as line 60? we should try to make everything consistent.

yichengq replied:
I tried it but thought it was too verbosal:
```
member 6e3bd23ae5f1eae0 is healthy [match: 102->104, next: 105->105, state: ProgressStateReplicate->ProgressStateReplicate]
```
So i dumped state change if it doesn't change.

yichengq commented:
i changed outputs based on comments and its current output is:
```
cluster is healthy: raft is making progress [commit index: 423->425]
leader is 924e2e83e93f2560
member 6e3bd23ae5f1eae0 is unhealthy because its match index is not catching up with commit index [match: 407->407, next: 408->408, state: ProgressStateProbe]
member 924e2e83e93f2560 is healthy [match: 423->425, next: 426->426, state: ProgressStateProbe]
member a8266ecf031671f3 is healthy [match: 423->425, next: 426->426, state: ProgressStateReplicate]
```
Look more informative?

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,43 @@ func handleClusterHealth(c *cli.Context) {
+		if pr1.Match <= pr0.Match {
+			var msg string
+			if pr0.State == pr1.State {
+				msg = fmt.Sprintf("member %s is unhealthy because its match index is not catching up with commit index [match: %v->%v, next: %v->%v, state: %v]\n", id, pr0.Match, pr1.Match, pr0.Next, pr1.Next, pr1.State)
```
can we make this consistent with the previous one? replace `because` with `:`?

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,43 @@ func handleClusterHealth(c *cli.Context) {
-			prints = append(prints, fmt.Sprintf("member %s is healthy\n", name))
+			var msg string
+			if pr0.State == pr1.State {
+				msg = fmt.Sprintf("member %s is healthy [match: %v->%v, next: %v->%v, state: %v]\n", id, pr0.Match, pr1.Match, pr1.Next, pr1.Next, pr1.State)
```
can we make this consistent with the previous one? member %s is healthy: raft is making progress [match ...] (we do not need next and state)

xiang90 commented:
LGTM after making everything consistent and short.

xiang90 commented:
Also please manually test the previous issue (hard shutdown a member has a reporting delay). I will give a test too soon. Thanks.

yichengq commented:
I have tested it on AWS and current way works well.

I addressed the comments and current log is as this:
```
cluster is healthy: raft is making progress [commit index: 848->850]
leader is 924e2e83e93f2560
member 6e3bd23ae5f1eae0 is unhealthy: its match index is not catching up with commit index [match: 839->839, next: 840->840, state: ProgressStateProbe]
member 924e2e83e93f2560 is healthy: raft is making progress [match: 848->850]
member a8266ecf031671f3 is healthy: raft is making progress [match: 848->850]
```

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -51,27 +50,37 @@ func handleClusterHealth(c *cli.Context) {
-		if fs1.Counts.Success <= fs0.Counts.Success {
-			prints = append(prints, fmt.Sprintf("member %s is unhealthy\n", name))
+		if pr1.Match <= pr0.Match {
+			var msg string
```
can we simply say: raft is not making progress and print out match as before?

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -99,13 +126,16 @@ func getLeaderStats(tr *http.Transport, endpoints []string) (string, *stats.Lead
+		return ep, vs.RaftStatus, nil
 	}
-	return "", nil, errors.New("no leader")
+	return "", raftStatus{}, errors.New("no leader")
```
well... if the given endpoint is not etcd 2.2 or not etcd, we will return no leader error. We can do better.

yichengq replied:
agree. I could do it in the following-up PR. I don't plan to do it in this PR because I don't change related logic and the error is not printed out.

xiang90 commented on etcdctl/command/cluster_health.go:
```diff
@@ -82,15 +85,33 @@ func handleClusterHealth(c *cli.Context) {
+}
+
+func getLeaderStatus(tr *http.Transport, endpoints []string) (string, raftStatus, error) {
 	// go-etcd does not support cluster stats, use http client for now
```
can we remove this irrelevant comment at line 107?

yichengq replied:
done.

xiang90 commented:
LGTM.
